File: frontend/invest/src/types/marketEvents.ts
export type MarketEventCategory =
  | "earnings"
  | "economic"
  | "disclosure"
  | "crypto_exchange_notice"
  | "crypto_protocol"
  | "tokenomics"
  | "regulatory";

export type MarketEventMarket = "us" | "kr" | "crypto" | "global";

export interface MarketEventValue {
  metric_name: string;
  period: string | null;
  actual: string | null;
  forecast: string | null;
  previous: string | null;
  revised_previous: string | null;
  unit: string | null;
  surprise: string | null;
  surprise_pct: string | null;
  released_at: string | null;
}

export interface MarketEvent {
  category: MarketEventCategory;
  market: MarketEventMarket;
  country: string | null;
  currency: string | null;
  symbol: string | null;
  company_name: string | null;
  title: string | null;
  event_date: string;
  release_time_utc: string | null;
  time_hint: string | null;
  importance: number | null;
  status: string;
  source: string;
  source_event_id: string | null;
  source_url: string | null;
  fiscal_year: number | null;
  fiscal_quarter: number | null;
  held: boolean | null;
  watched: boolean | null;
  values: MarketEventValue[];
}

export interface MarketEventsDayResponse {
  date: string;
  events: MarketEvent[];
}

export interface FetchMarketEventsTodayParams {
  category?: MarketEventCategory;
  market?: MarketEventMarket;
  source?: string;
  /** ISO date — when omitted the backend defaults to today (server clock). */
  onDate?: string;
}

File: app/schemas/market_events.py
"""Pydantic response schemas for market events (ROB-128)."""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class MarketEventValueResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    metric_name: str
    period: str | None = None
    actual: Decimal | None = None
    forecast: Decimal | None = None
    previous: Decimal | None = None
    revised_previous: Decimal | None = None
    unit: str | None = None
    surprise: Decimal | None = None
    surprise_pct: Decimal | None = None
    released_at: datetime | None = None


class MarketEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    category: str
    market: str
    country: str | None = None
    currency: str | None = None
    symbol: str | None = None
    company_name: str | None = None
    title: str | None = None
    event_date: date
    release_time_utc: datetime | None = None
    time_hint: str | None = None
    importance: int | None = None
    status: str = "scheduled"
    source: str
    source_event_id: str | None = None
    source_url: str | None = None
    fiscal_year: int | None = None
    fiscal_quarter: int | None = None

    held: bool | None = None
    watched: bool | None = None

    values: list[MarketEventValueResponse] = Field(default_factory=list)


class MarketEventsDayResponse(BaseModel):
    date: date
    events: list[MarketEventResponse] = Field(default_factory=list)


class MarketEventsRangeResponse(BaseModel):
    from_date: date
    to_date: date
    count: int
    events: list[MarketEventResponse] = Field(default_factory=list)


class IngestionPartitionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    source: str
    category: str
    market: str
    partition_date: date
    status: str
    event_count: int
    started_at: datetime | None = None
    finished_at: datetime | None = None
    last_error: str | None = None
    retry_count: int


class IngestionRunResult(BaseModel):
    source: str
    category: str
    market: str
    partition_date: date
    status: str
    event_count: int
    error: str | None = None
